aquadoggo/src/replication/service.rs

```diff
@@ -0,0 +1,275 @@
+        tx: &ServiceSender,
+        local_peer_id: PeerId,
+        target_set: TargetSet,
+    ) -> Self {
+        let ingest = SyncIngest::new(schema_provider.clone(), tx.clone());
+        let sync_manager = SyncManager::new(store.clone(), ingest, local_peer_id);
+
+        Self {
+            peers: HashMap::new(),
+            sync_manager,
+            tx: tx.clone(),
+            rx: tx.subscribe(),
+            target_set,
+        }
+    }
+
+    async fn on_connection_established(&mut self, peer_id: PeerId) {
+        info!("Connection established with peer: {}", peer_id);
+        if self
+            .peers
+            .insert(peer_id, PeerStatus::new(&peer_id))
+            .is_some()
+        {
+            warn!("Duplicate established connection encountered");
+        }
+
+        self.update_sessions().await;
+    }
+
+    async fn on_connection_closed(&mut self, peer_id: PeerId) {
+        // Clear running replication sessions from sync manager
+        info!("Connection closed: remove sessions with peer: {}", peer_id);
+        self.sync_manager.remove_sessions(&peer_id);
+
+        // Remove peer from our connections table
+        if self.peers.remove(&peer_id).is_none() {
+            warn!("Tried to remove unknown connection");
+        }
+
+        self.update_sessions().await;
+    }
+
+    async fn on_replication_message(&mut self, peer_id: PeerId, message: SyncMessage) {
+        trace!("Received SyncMessage: {:?}", message);
+        match self.sync_manager.handle_message(&peer_id, &message).await {
+            Ok(result) => {
+                for message in result.messages {
+                    self.send_service_message(ServiceMessage::SentReplicationMessage(
+                        peer_id, message,
+                    ));
+                }
+
+                if result.is_done {
+                    self.on_replication_finished(peer_id).await;
+                }
+            }
+            Err(err) => {
+                self.on_replication_error(peer_id, err).await;
+            }
+        }
+    }
+
+    async fn on_replication_finished(&mut self, peer_id: PeerId) {
+        info!("Finished replication with peer {}", peer_id);
+
+        match self.peers.get_mut(&peer_id) {
+            Some(status) => {
+                status.successful_count += 1;
+            }
+            None => {
+                panic!("Tried to access unknown peer");
+            }
+        }
+
+        self.update_sessions().await;
+    }
+
+    async fn on_replication_error(&mut self, peer_id: PeerId, error: ReplicationError) {
+        info!("Replication with peer {} failed: {}", peer_id, error);
+
+        match self.peers.get_mut(&peer_id) {
+            Some(status) => {
+                status.failed_count += 1;
+            }
+            None => {
+                panic!("Tried to access unknown peer");
+            }
+        }
+
+        // @TODO: SyncManager should remove session internally on critical errors
```

Comment:
Which cases does this message refer to @adzialocha ?

Reply:
We haven't specified that yet but I assume all "critical" errors should terminate replication. I guess it also depends how strict we want to be.